Add Bearer prefix to Authentication header when using bearer security scheme

androa opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

**What**: Add `Bearer` prefix to the `Authorization` header when API is using `Bearer` based authentication.

<!-- Why are these changes necessary? -->

**Why**: When setting up an authentication scheme using `bearer` as `scheme` the requests made when contract testing should contain a `Bearer` prefix in the generated `Authentication` header.

Many frameworks, `ktor` included, requires the `Bearer` prefix to be there to correctly delegate to the specified auth handler. Without it `ktor` automatically rejects the request with a `401 Unauthorized` when the `Authorization` header simply is `Authorization: WFPSL`.

<!-- How were these changes implemented? -->

**How**: Add a `Bearer` prefix to the `Authorization` header.

<!-- Have you done all of these things?  -->

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [ ] Documentation added to the README.md OR link to PR on https://github.com/specmatic/specmatic-documentation
- [x] Tests
- [ ] Sonar Quality Gate

androa committed:
Add Bearer prefix to Authentication header when using bearer security scheme

androa committed:
Ensure that authentication headers starts with Bearer when stubbing

androa commented on core/src/main/kotlin/in/specmatic/conversions/BearerSecurityScheme.kt:
```diff
@@ -9,7 +9,9 @@ import org.apache.http.HttpHeaders.AUTHORIZATION
     override fun matches(httpRequest: HttpRequest): Boolean {
-        return httpRequest.headers.containsKey(AUTHORIZATION)
+        return httpRequest.headers.let {
+            it.containsKey(AUTHORIZATION) && it[AUTHORIZATION]!!.lowercase().startsWith("bearer")
```
@joelrosario Would that be something like this? It is rather awkward to not get a better error than the generic `No auth params were found in the request`, but I'm not quite sure the best way to change the signature of `matches()` in a way it can create different failure responses.

joelrosario replied:
@androa yes, and thanks, that change looks good!

But you're also right about the generic message not being too helpful.

We'd use Specmatic's Result class to provide more detail.

This means, we'd need to change the return type of the `matches` method from Boolean to Result.

I believe there should not be much impact as this method is not referred to in many places, which include:
1. the other classes implementing OpenAPISecurityScheme
2. `HttpRequestPattern.matchSecurityScheme`... we'd need to modify a line of code in that method to match the new signature.

androa replied:
Made an attempt of refactoring it into returning the Result class now.

androa committed:
Change return type of OpenAPISecurityScheme::matches() from boolean to Result

This allows more verbose feedback on why the authentication mechanism did not match the request

joelrosario reviewed (approved):
Looks good to me! Thanks @androa for your patience and for the PR, really appreciate it.
